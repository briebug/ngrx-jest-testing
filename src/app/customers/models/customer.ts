export interface Customer {
  id: number;
  name: string;
}
