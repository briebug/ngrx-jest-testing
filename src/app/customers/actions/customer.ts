import { Action } from '@ngrx/store';
import { Customer } from '../models/customer';

export enum CustomerActionTypes {
  Load = '[Customer] Load',
  LoadSuccess = '[Customer] Load Success',
  LoadFail = '[Customer] Load Fail',
  Select = '[Customer] Select'
}

export class Load implements Action {
  readonly type = CustomerActionTypes.Load;
}

export class LoadSuccess implements Action {
  readonly type = CustomerActionTypes.LoadSuccess;

  constructor(public payload: Customer[]) {}
}

export class LoadFail implements Action {
  readonly type = CustomerActionTypes.LoadFail;

  constructor(public error: any) {}
}

export class Select implements Action {
  readonly type = CustomerActionTypes.Select;

  constructor(public payload: Customer) {}
}

export type CustomerActions = Load | LoadSuccess | LoadFail | Select;
