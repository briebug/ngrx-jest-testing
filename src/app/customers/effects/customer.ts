import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';

import {
  CustomerActionTypes,
  Load,
  LoadSuccess,
  LoadFail
} from '../actions/customer';

import { Customer } from '../models/customer';
import { map, switchMap, catchError } from 'rxjs/operators';
import { CustomersService } from '../services/customers';

@Injectable()
export class CustomerEffects {
  @Effect()
  load$: Observable<Action> = this.actions$.pipe(
    ofType<Load>(CustomerActionTypes.Load),
    switchMap(query => {
      return this.customersService
        .load()
        .pipe(
          map((customers: Customer[]) => new LoadSuccess(customers)),
          catchError(err => of(new LoadFail(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private customersService: CustomersService
  ) {}
}
