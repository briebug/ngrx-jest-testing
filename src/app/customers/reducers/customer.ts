import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { Customer } from '../models/customer';
import { CustomerActions, CustomerActionTypes } from '../actions/customer';

export interface State extends EntityState<Customer> {
  selectedCustomerId: number | null;
  loading: boolean;
  error: any;
}

export const adapter: EntityAdapter<Customer> = createEntityAdapter<Customer>();

export const initialState: State = adapter.getInitialState({
  selectedCustomerId: null,
  loading: false,
  error: null
});

export function reducer(state = initialState, action: CustomerActions): State {
  switch (action.type) {
    case CustomerActionTypes.Load:
      return { ...state, loading: true, selectedCustomerId: null };

    case CustomerActionTypes.LoadSuccess: {
      return { ...adapter.addMany(action.payload, state) };
    }

    case CustomerActionTypes.LoadFail: {
      return { ...state, error: { message: 'Error loading customers' } };
    }

    case CustomerActionTypes.Select: {
      return {
        ...state,
        selectedCustomerId: action.payload.id
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedCustomerId;
export const getLoading = (state: State) => state.loading;
export const getError = (state: State) => state.error;
