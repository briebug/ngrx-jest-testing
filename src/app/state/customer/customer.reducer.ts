import { createEntityAdapter, EntityAdapter, EntityState } from '@ngrx/entity';

import { Customer } from './customer.model';
import { CustomerActions, CustomerActionTypes } from './customer.actions';

export interface State extends EntityState<Customer> {
  selectedCustomerId: number;
  loading: boolean;
  error: string;
}

export const adapter: EntityAdapter<Customer> = createEntityAdapter<Customer>();

export const initialState: State = adapter.getInitialState({
  selectedCustomerId: null,
  loading: false,
  error: null
});

export function reducer(state = initialState, action: CustomerActions): State {
  switch (action.type) {
    case CustomerActionTypes.Load:
      return { ...state, loading: true, selectedCustomerId: null };

    case CustomerActionTypes.LoadSuccess: {
      return { ...adapter.addMany(action.payload, state) };
    }

    case CustomerActionTypes.LoadFail: {
      console.log(action.error);
      return { ...state, error: 'Error loading customers' };
    }

    case CustomerActionTypes.Select: {
      return {
        ...state,
        selectedCustomerId: action.payload.id
      };
    }

    default: {
      return state;
    }
  }
}

export const getSelectedId = (state: State) => state.selectedCustomerId;
export const getLoading = (state: State) => state.loading;
export const getError = (state: State) => state.error;
