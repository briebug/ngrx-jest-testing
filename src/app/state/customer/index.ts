import { createSelector, createFeatureSelector } from '@ngrx/store';

import * as fromCustomers from './customer.reducer';
import { State as CustomersState } from './customer.reducer';

export const reducers = {
  customers: fromCustomers.reducer
};


export const getCustomersState = createFeatureSelector<CustomersState>(
  'customer'
);

export const {
  selectAll: getAllCustomers,
  selectEntities: getCustomerEntities,
  selectIds: getCustomerIds,
  selectTotal: getTotalCustomers
} = fromCustomers.adapter.getSelectors(getCustomersState);

export const getSelectedCustomerId = createSelector(
  getCustomersState,
  fromCustomers.getSelectedId
);

export const getSelectedCustomer = createSelector(
  getCustomerEntities,
  fromCustomers.getSelectedId,
  (entities, id) => entities[id]
);

export const getLoading = createSelector(
  getCustomersState,
  fromCustomers.getLoading
);

export const getError = createSelector(
  getCustomersState,
  fromCustomers.getError
);
