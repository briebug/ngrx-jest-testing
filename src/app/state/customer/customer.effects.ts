import { Injectable, InjectionToken, Optional, Inject } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';
import { map, switchMap, catchError } from 'rxjs/operators';

import {
  CustomerActionTypes,
  Load,
  LoadSuccess,
  LoadFail
} from './customer.actions';

import { Customer } from './customer.model';
import { CustomersService } from '@core/services/customer.service';

@Injectable()
export class CustomerEffects {
  @Effect()
  load$: Observable<Action> = this.actions$.pipe(
    ofType<Load>(CustomerActionTypes.Load),
    switchMap(query => {
      return this.customersService
        .load()
        .pipe(
          map((customers: Customer[]) => new LoadSuccess(customers)),
          catchError(err => of(new LoadFail(err)))
        );
    })
  );

  constructor(
    private actions$: Actions,
    private customersService: CustomersService
  ) {}
}
