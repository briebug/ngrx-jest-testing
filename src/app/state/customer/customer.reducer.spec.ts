import { reducer } from './customer.reducer';
import * as fromCustomers from './customer.reducer';
import * as selectors from '.';
import { Load, LoadSuccess, LoadFail, Select } from './customer.actions';
import { Customer } from './customer.model';

describe('CustomersReducer', () => {
  const initialState: fromCustomers.State = {
    ids: [],
    entities: {},
    loading: false,
    selectedCustomerId: null,
    error: null
  },
    customers = [
      { id: 1, name: 'test1' },
      { id: 2, name: 'test2' },
      { id: 3, name: 'test3' }
    ],
    customer = customers[0];

  describe('undefined action', () => {
    it('should return the default state', () => {
      const result = reducer(undefined, {} as any);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Load Customers Action', () => {
    test('should set loading to true', () => {
      const action = new Load(),
        result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
      expect(fromCustomers.getLoading(result)).toBe(true);
    });
  });

  describe('Load Customers Success Action', () => {
    test('should add customers to state', () => {
      const action = new LoadSuccess(customers),
        result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
    });
  });

  describe('Load Customers Fail Action', () => {
    test('should add error to state', () => {
      const action = new LoadFail({ status: 1 }),
        result = reducer(initialState, action),
        error = { message: 'Error loading customers' };

      expect(result).toMatchSnapshot();
      expect(fromCustomers.getError(result)).toEqual(error);
    });
  });

  describe('Customer Select Action', () => {
    test('should set selectedCustomerId', () => {
      const action = new Select(customer),
        result = reducer(initialState, action);

      expect(result).toMatchSnapshot();
      expect(fromCustomers.getSelectedId(result)).toEqual(customer.id);
    });
  });

  describe('GetSelectedCustomer selector', () => {
    test('should select customer', () => {
      const action = new LoadSuccess(customers);
      let result = reducer(initialState, action);

      result = reducer(result, new Select(customer));

      expect(selectors.getSelectedCustomer.projector(result.entities, result.selectedCustomerId)).toMatchSnapshot();
    });
  });
});
