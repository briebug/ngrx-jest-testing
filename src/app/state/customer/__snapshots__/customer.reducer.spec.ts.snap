// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomersReducer Customer Select Action should set selectedCustomerId 1`] = `
Object {
  "entities": Object {},
  "error": null,
  "ids": Array [],
  "loading": false,
  "selectedCustomerId": 1,
}
`;

exports[`CustomersReducer GetSelectedCustomer selector should select customer 1`] = `
Object {
  "id": 1,
  "name": "test1",
}
`;

exports[`CustomersReducer Load Customers Action should set loading to true 1`] = `
Object {
  "entities": Object {},
  "error": null,
  "ids": Array [],
  "loading": true,
  "selectedCustomerId": null,
}
`;

exports[`CustomersReducer Load Customers Fail Action should add error to state 1`] = `
Object {
  "entities": Object {},
  "error": "Error loading customers",
  "ids": Array [],
  "loading": false,
  "selectedCustomerId": null,
}
`;

exports[`CustomersReducer Load Customers Success Action should add customers to state 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "id": 1,
      "name": "test1",
    },
    "2": Object {
      "id": 2,
      "name": "test2",
    },
    "3": Object {
      "id": 3,
      "name": "test3",
    },
  },
  "error": null,
  "ids": Array [
    1,
    2,
    3,
  ],
  "loading": false,
  "selectedCustomerId": null,
}
`;

exports[`CustomersReducer undefined action should return the default state 1`] = `
Object {
  "entities": Object {},
  "error": null,
  "ids": Array [],
  "loading": false,
  "selectedCustomerId": null,
}
`;
